AppKit/CPWindow/CPWindow.j

```diff
@@ -3844,6 +3819,393 @@ var interpolate = function(fromValue, toValue, progress)
 
 @end
 
+@implementation CPWindow (TrackingAreaAdditions)
+
+- (void)_addTrackingAreaView:(CPView)aView
+{
+    var trackingAreas = [aView trackingAreas];
+    
+    for (var i = 0; i < trackingAreas.length; i++)
+        [self _addTrackingArea:trackingAreas[i]];
+}
+
+- (void)_removeTrackingAreaView:(CPView)aView
+{
+    var trackingAreas = [aView trackingAreas];
+    
+    for (var i = 0; i < trackingAreas.length; i++)
+        [self _removeTrackingArea:trackingAreas[i]];
+}
+
+- (void)_addTrackingArea:(CPTrackingArea)trackingArea
+{
+    var trackingAreaView = [trackingArea view];
+    
+    if (![_trackingAreaViews containsObject:trackingAreaView])
+        [_trackingAreaViews addObject:trackingAreaView];
+    
+    // If CPTrackingAssumeInside option is set, put the tracking area in the _mouseEnteredStack
+    
+    if ([trackingArea options] & CPTrackingAssumeInside)
+        [_mouseEnteredStack addObject:trackingArea];
+}
+
+- (void)_removeTrackingArea:(CPTrackingArea)trackingArea
+{
+    // If mouse is in the tracking area, we remove it from the stack to avoid to fire a future mouseExited event
+    
+    [_mouseEnteredStack removeObjectIdenticalTo:trackingArea];
+    
+    var trackingAreaView = [trackingArea view];
+    
+    [_trackingAreaViews removeObjectIdenticalTo:trackingAreaView];
+}
+
+- (void)_handleTrackingAreaEvent:(CPEvent)anEvent
```

Comment:
@Dogild This is a **perfect** example of micro-optimizations that effectively have zero effect on performance. As I said earlier, humans are very bad at figuring out where performance bottlenecks are. I did some tests on using a property vs. using a variable:

- Safari 9: variable is 30% faster
- Chrome: same
- FireFox: same
- IE 11: variable is 2x faster

These numbers are meaningless unless we look at the broader picture. 30% faster than what? What is the usage share of the browser?

30% faster on Safari sounds great until you look closer. It turns out that in that test, Safari took about 3 **nanoseconds** (that's **billionths** of a second) per loop iteration. So 1 **million** loop iterations takes about 3 milliseconds (thousands of a second), and 30% faster than that is 2 millseconds, and thus even with a million loop iterations we save only 1 millisecond.

In any case, Chrome has the largest share of the browser market by far. And on Chrome their is no performance gain. So I'll say it again: it really isn't worth the extra code and loss of readability in order to save a millisecond, especially when Javascript engines are continually getting faster.